### PR TITLE
optimize and cleanup bootstrap source

### PR DESCRIPTION
**Pull request by onur-ozkan**
I suggest reviewing this commit by commit.

### CODE REVIEW

**Comment by onur-ozkan** [comment]
Architectural changes will be done in a separate PR to avoid complicating the review process for this PR.

@rustbot ready

**Review by albertlarsan68 — APPROVED** [review]

File: src/bootstrap/bin/rustc.rs [hunk 1]
```diff
@@ -15,14 +15,14 @@
 //! switching compilers for the bootstrap and for build scripts will probably
 //! never get replaced.
 
-include!("../dylib_util.rs");
```
Comment (bjorn3) [inline src/bootstrap/bin/rustc.rs:18]: I introduced `dylib_util.rs` to reduce compilation time by avoiding linking of libbootstrap and all it's dependencies when compiling the auxiliary helper executables.

File: src/bootstrap/CHANGELOG.md [hunk 2]
```diff
@@ -1,71 +0,0 @@
-# Changelog
```
Comment (bjorn3) [inline src/bootstrap/CHANGELOG.md:1]: If you are deleting the changelog, maybe also delete the `changelog-seen` key in config.toml? This key and the changelog was used when we made significant changes to the defaults of bootstrap affecting everyone's workflows.

**Comment by onur-ozkan** [comment]
> I introduced dylib_util.rs to reduce compilation time by avoiding linking of libbootstrap and all it's dependencies when compiling the auxiliary helper executables.

Whill check it.

> If you are deleting the changelog, maybe also delete the changelog-seen key in config.toml? This key and the changelog was used when we made significant changes to the defaults of bootstrap affecting everyone's workflows.

It's still useful, but can make more sense if we find better name and use the PR links instead of versions(similar to llvm).

@bors r-

File: src/bootstrap/bin/rustc.rs (hunk 1 above)
Comment (onur-ozkan) [reply src/bootstrap/bin/rustc.rs:18]: I am quite impressed from this hack :D Thank you for reminding.

**Comment by onur-ozkan** [comment]
Changelog related commits will be cherry-picked to another PR that I am planning to refactor this `changelog-seen` key as it's not really functioning as expected(e.g., we did many changes on the build flows, but never updated the key or the Changelog files).

Until then, @rustbot block

**Comment by onur-ozkan** [comment]
Changelog related commits have been moved to https://github.com/rust-lang/rust/pull/115898.

No changes since the r+ from @albertlarsan68, just reduced some commits. @bors r=albertlarsan68

**Pull request merged**